SofaKernel/modules/SofaSimulationCore/src/sofa/simulation/BaseMechanicalVisitor.cpp

```diff
@@ -46,199 +46,123 @@ namespace sofa::simulation
     {
-        res = for_each_r(this, ctx, node->interactionForceField, &BaseMechanicalVisitor::fwdInteractionForceField);
+        return RESULT_PRUNE;
     }
 
+    if (for_each(this, ctx, node->forceField, &BaseMechanicalVisitor::fwdForceField, fwdVisitorType) == RESULT_PRUNE)
+    {
+        return RESULT_PRUNE;
+    }
 
-    if (res != RESULT_PRUNE)
+    if (for_each(this, ctx, node->interactionForceField, &BaseMechanicalVisitor::fwdInteractionForceField, fwdVisitorType) == RESULT_PRUNE)
     {
-        res = for_each_r(this, ctx, node->projectiveConstraintSet, &BaseMechanicalVisitor::fwdProjectiveConstraintSet);
+        return RESULT_PRUNE;
     }
 
-    if (res != RESULT_PRUNE)
+    if (for_each(this, ctx, node->projectiveConstraintSet, &BaseMechanicalVisitor::fwdProjectiveConstraintSet, fwdVisitorType) == RESULT_PRUNE)
     {
-        res = for_each_r(this, ctx, node->constraintSet, &BaseMechanicalVisitor::fwdConstraintSet);
+        return RESULT_PRUNE;
     }
 
+    if (for_each(this, ctx, node->constraintSet, &BaseMechanicalVisitor::fwdConstraintSet, fwdVisitorType) == RESULT_PRUNE)
+    {
+        return RESULT_PRUNE;
+    }
 
-    return res;
+    return RESULT_CONTINUE;
 }
 
 
-void BaseMechanicalVisitor::processNodeBottomUp(simulation::Node* node, VisitorContext* ctx)
+void BaseMechanicalVisitor::processNodeBottomUp(simulation::Node *node, VisitorContext *ctx)
 {
-    for_each(this, ctx, node->projectiveConstraintSet, &BaseMechanicalVisitor::bwdProjectiveConstraintSet);
-    for_each(this, ctx, node->constraintSet, &BaseMechanicalVisitor::bwdConstraintSet);
-    for_each(this, ctx, node->constraintSolver, &BaseMechanicalVisitor::bwdConstraintSolver);
+    for_each(this, ctx, node->projectiveConstraintSet, &BaseMechanicalVisitor::bwdProjectiveConstraintSet, bwdVisitorType);
+    for_each(this, ctx, node->constraintSet, &BaseMechanicalVisitor::bwdConstraintSet, bwdVisitorType);
+    for_each(this, ctx, node->constraintSolver, &BaseMechanicalVisitor::bwdConstraintSolver, bwdVisitorType);
 
     if (node->mechanicalState != nullptr)
     {
         if (node->mechanicalMapping != nullptr)
         {
             if (!stopAtMechanicalMapping(node, node->mechanicalMapping))
             {
-                if(testTags(node->mechanicalState))
+                if (testTags(node->mechanicalState))
                 {
-                    ctime_t t = begin(node, node->mechanicalState, "bwd");
-                    this->bwdMappedMechanicalState(ctx, node->mechanicalState);
-                    end(node, node->mechanicalState, t);
-                    t = begin(node, node->mechanicalMapping, "bwd");
-                    this->bwdMechanicalMapping(ctx, node->mechanicalMapping);
-                    end(node, node->mechanicalMapping, t);
+                    runVisitorTask(this, ctx, &BaseMechanicalVisitor::bwdMappedMechanicalState, &*node->mechanicalState, bwdVisitorType);
```

Comment:
Maybe move the first runVisitorTask on mappedMS out of the tag test (as runVisitor will also test the tags)...
but this comment is mostly to show I did a thorough review 😁

Reply:
In any case, the tag test will be performed twice

Reply:
true that